tutorials/generative/anomaly_detection/anomaly_detection_with_transformers.py

```diff
@@ -420,33 +422,153 @@
+
+# %% [markdown]
+# Get the log-likelihood and convert into a mask of the 5% lowest-likelihood tokens
 
 # %%
+log_likelihood = inferer.get_likelihood(
+    inputs=image_corrupted[None, ...].to(device),
+    vqvae_model=vqvae_model,
+    transformer_model=transformer_model,
+    ordering=ordering,
+)
+plt.subplot(1, 2, 1)
+plt.imshow(log_likelihood.cpu()[0, ...], vmin=0.6, vmax=1)
+plt.axis("off")
+plt.title("Log-likelihood")
+plt.subplot(1, 2, 2)
+mask = log_likelihood.cpu()[0, ...] < torch.quantile(log_likelihood, 0.05).item()
+plt.imshow(mask)
+plt.axis("off")
+plt.title("Healing mask")
+plt.show()
+
+# %% [markdown]
+# Use this mask and the trained transformer to 'heal' the sequence
+
+# %%
+# flatten the mask
+mask_flattened = mask.reshape(-1)
+mask_flattened = mask_flattened[ordering.get_sequence_ordering()]
+
+latent = vqvae_model.index_quantize(image_corrupted[None, ...].to(device))
+latent = latent.reshape(latent.shape[0], -1)
+latent = latent[:, ordering.get_sequence_ordering()]
+latent = F.pad(latent, (1, 0), "constant", vqvae_model.num_embeddings)
+latent = latent.long()
+latent_healed = latent.clone()
+
+# heal the sequence
+# loop over tokens
+for i in range(1, latent.shape[1]):
+    if mask_flattened[i - 1]:
+        # if token is low probability, replace with tranformer's most likely token
+        logits = transformer_model(latent_healed[:, :i])
+        probs = F.softmax(logits, dim=-1)
+        # don't sample beginning of sequence token
+        probs[:, :, vqvae_model.num_embeddings] = 0
+        index = torch.argmax(probs[0, -1, :])
+        latent_healed[:, i] = index
+
+
+# reconstruct
+latent_healed = latent_healed[:, 1:]
+latent_healed = latent_healed[:, ordering.get_revert_sequence_ordering()]
+latent_healed = latent_healed.reshape((16, 16))
+
+image_healed = vqvae_model.decode_samples(latent_healed[None, ...]).cpu().detach()
+plt.imshow(image_healed[0, 0, ...], cmap="gray")
+plt.axis("off")
+plt.title("Healed image")
+plt.show()
+
+# %% [markdown]
+# ## Create anomaly maps
+
+# %%
+# Get a naive anomaly map using the difference
+difference_map = torch.abs(image_healed[0, 0, ...] - image_corrupted[0, ...])
+
+# Further mask with the healing mask
+resizer = torch.nn.Upsample(size=(64, 64), mode="nearest")
+mask_upsampled = resizer(mask[None, None, ...].float()).int()
+
+fig, ax = plt.subplots(1, 4, figsize=(14, 8))
+plt.subplot(1, 4, 1)
+plt.imshow(image_clean[0, ...], cmap="gray")
+plt.axis("off")
+plt.title("Clean image")
+image_corrupted = image_clean.clone()
+image_corrupted[0, 25:40, 40:50] = 1
+plt.subplot(1, 4, 2)
+plt.imshow(image_corrupted[0, ...], cmap="gray")
+plt.axis("off")
+plt.title("Corrupted image")
+plt.subplot(1, 4, 3)
+plt.imshow(image_corrupted[0, ...] - image_clean[0, ...], cmap="gray")
+plt.axis("off")
+plt.title("Ground-Truth anomaly mask")
+plt.subplot(1, 4, 4)
+plt.imshow(mask_upsampled[0, 0, ...] * difference_map, cmap="gray")
+plt.axis("off")
+plt.title("Predicted anomaly mask")
+plt.show()
+
```

Comment:
I guess it is necessary to run `./runtests.sh --autofix`. here When I try it, it changes this last lane.